### PR TITLE
ascanrules: SSTI blind using timingutils

### PR DESCRIPTION
**Pull request by itsdivyanshjain**
## Overview
Briefly describe the purpose, goals, and changes or improvements made in this pull request.

## Related Issues
Specify any related issues or pull requests by linking to them.

## Checklist
- [ ] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [x] Write tests
- [ ] Check code coverage
- [ ] Sign-off commits
- [ ] Squash commits
- [x] Use a descriptive title

For more details, please refer to the [developer rules and guidelines](https://www.zaproxy.org/docs/developer/dev-rules-and-guidelines/).

### CODE REVIEW

**Comment by itsdivyanshjain** [comment]
before reviewing this , first merge (https://github.com/zaproxy/zap-extensions/pull/5161)

**Comment by thc202** [comment]
Why not remove the unrelated commits?

File: addOns/ascanrules/CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Improved the following scan rules by using time-based linear regression tests:
   - SQL Injection - Hypersonic SQL
   - SQL Injection - MsSQL
+  - SQL Injection - MySQL
+  - SSTI - Blind
```
Comment (thc202) [inline addOns/ascanrules/CHANGELOG.md:13]: `Server Side Template Injection (Blind)`

**Comment by itsdivyanshjain** [comment]
@thc202 done

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRule.java [hunk 2]
```diff
@@ -181,61 +215,65 @@ private void timeBasedTests(String paramName, String[] commandExecPayloads) {
...
+    private boolean checkIfCausesTimeDelay(String paramName, String payloadFormat) {
+        AtomicReference<HttpMessage> message = new AtomicReference<>();
+        AtomicReference<String> attack = new AtomicReference<>();
+        TimingUtils.RequestSender requestSender =
+                x -> {
+                    HttpMessage msg = getNewMsg();
+                    message.compareAndSet(null, msg);
+
+                    String finalPayload =
+                            payloadFormat.replace(SECONDS_PLACEHOLDER, String.valueOf(x));
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRule.java:227]:
```suggestion
                            payloadFormat.replace(SECONDS_PLACEHOLDER, Integer.toString((int) x));
```

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRule.java [hunk 3]
```diff
@@ -181,61 +215,65 @@ private void timeBasedTests(String paramName, String[] commandExecPayloads) {
...
+
+                    setParameter(msg, paramName, finalPayload);
+                    LOGGER.debug("Testing [{}] = [{}]", paramName, finalPayload);
+
+                    attack.compareAndSet(null, finalPayload);
+                    sendAndReceive(msg, false);
+                    return msg.getTimeElapsedMillis() / 1000.0;
+                };
+        // end of timing baseline check
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRule.java:236]:
```suggestion
```

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRuleUnitTest.java [hunk 4]
```diff
@@ -41,4 +54,78 @@ protected int getRecommendMaxNumberMessagesPerParam(Plugin.AttackStrength streng
                 return recommendMax;
         }
     }
+
+    @Test
+    void shouldAlertIfSleepTimesGetLonger() throws Exception {
+        String test = "/shouldReportSqlTimingIssue";
+        Pattern sleepPattern = Pattern.compile("#\\{%x\\(sleep\\s+(\\d+)\\.\\d+\\)\\}");
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRuleUnitTest.java:61]:
```suggestion
        Pattern sleepPattern = Pattern.compile("#\\{%x\\(sleep\\s+(\\d+)\\)\\}");
```

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRuleUnitTest.java [hunk 5]
```diff
@@ -41,4 +54,78 @@ protected int getRecommendMaxNumberMessagesPerParam(Plugin.AttackStrength streng
...
+        HttpMessage msg = this.getHttpMessage(test + "?name=test");
+
+        this.rule.init(msg, this.parent);
+        this.rule.setSleepInSeconds(2);
+
+        this.rule.scan();
+
+        assertThat(alertsRaised.size(), equalTo(1));
+        assertThat(alertsRaised.get(0).getParam(), equalTo("name"));
+        assertThat(alertsRaised.get(0).getAttack(), equalTo("#{%x(sleep 2.0)}"));
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRuleUnitTest.java:96]:
```suggestion
        assertThat(alertsRaised.get(0).getAttack(), equalTo("#{%x(sleep 2)}"));
```

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRuleUnitTest.java [hunk 6]
```diff
@@ -19,7 +19,20 @@
  */
 package org.zaproxy.zap.extension.ascanrules;
 
+import static fi.iki.elonen.NanoHTTPD.newFixedLengthResponse;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+import fi.iki.elonen.NanoHTTPD.IHTTPSession;
+import fi.iki.elonen.NanoHTTPD.Response;
+// import static org.hamcrest.Matchers.is;
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRuleUnitTest.java:28]:
```suggestion
```

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRuleUnitTest.java [hunk 7]
```diff
@@ -41,4 +54,78 @@ protected int getRecommendMaxNumberMessagesPerParam(Plugin.AttackStrength streng
                 return recommendMax;
         }
     }
+
+    @Test
+    void shouldAlertIfSleepTimesGetLonger() throws Exception {
+        String test = "/shouldReportSqlTimingIssue";
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SstiBlindScanRuleUnitTest.java:60]:
```suggestion
        String test = "/shouldAlertIfSleepTimesGetLonger";
```

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by kingthorin — APPROVED** [review]

**Pull request merged**